extensions/rv32im/circuit/src/hintstore/mod.rs

```diff
@@ -391,51 +450,151 @@ where
             let data: [u8; RV32_REGISTER_NUM_LIMBS] =
                 data_f.map(|byte| byte.as_canonical_u32() as u8);
 
-            let mem_ptr_word = mem_ptr + (RV32_REGISTER_NUM_LIMBS * word_index) as u32;
+            record.var[idx].data = data;
 
-            row.data = data_f;
             tracing_write(
                 state.memory,
                 RV32_MEMORY_AS,
-                mem_ptr_word,
+                record.inner.mem_ptr + (RV32_REGISTER_NUM_LIMBS * idx) as u32,
                 &data,
-                &mut row.write_aux,
+                &mut record.var[idx].data_write_aux.prev_timestamp,
+                &mut record.var[idx].data_write_aux.prev_data,
             );
-
-            row.rem_words_limbs = decompose(num_words - word_index as u32);
-            row.mem_ptr_limbs = decompose(mem_ptr_word);
         }
 
         *state.pc = state.pc.wrapping_add(DEFAULT_PC_STEP);
 
-        *trace_offset += (num_words as usize) * width;
-
         Ok(())
     }
-
-    fn fill_trace_row(&self, mem_helper: &MemoryAuxColsFactory<F>, row_slice: &mut [F]) {
-        let row: &mut Rv32HintStoreCols<F> = row_slice.borrow_mut();
-
-        let mut timestamp = row.from_state.timestamp.as_canonical_u32();
-
-        if row.is_single.is_one() || row.is_buffer_start.is_one() {
-            mem_helper.fill_from_prev(timestamp, row.mem_ptr_aux_cols.as_mut());
+}
+impl<F: PrimeField32, CTX> TraceFiller<F, CTX> for Rv32HintStoreStep<F> {
+    fn fill_trace(
+        &self,
+        mem_helper: &MemoryAuxColsFactory<F>,
+        trace: &mut RowMajorMatrix<F>,
+        rows_used: usize,
+    ) where
+        Self: Send + Sync,
+        F: Send + Sync + Clone,
+    {
+        if rows_used == 0 {
+            return;
         }
-        timestamp += 1;
 
-        if row.is_buffer_start.is_one() {
-            mem_helper.fill_from_prev(timestamp, row.num_words_aux_cols.as_mut());
+        let mut sizes = Vec::with_capacity(rows_used);
```

Comment:
I feel like you don't need this? you can just do one loop through and move the pointer, for each chunk still assign in reverse from bottom of the chunk